client/backend/src/main/kotlin/com/microsoft/tfs/Main.kt

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root.
+
+package com.microsoft.tfs
+
+import com.jetbrains.rd.framework.*
+import com.jetbrains.rd.util.Logger
+import com.jetbrains.rd.util.info
+import com.jetbrains.rd.util.lifetime.Lifetime
+import com.jetbrains.rd.util.lifetime.LifetimeDefinition
+import com.jetbrains.rd.util.lifetime.isAlive
+import com.jetbrains.rd.util.threading.SingleThreadScheduler
+import com.microsoft.tfs.core.httpclient.UsernamePasswordCredentials
+import com.microsoft.tfs.model.host.TfsCollection
+import com.microsoft.tfs.model.host.TfsCollectionDefinition
+import com.microsoft.tfs.model.host.TfsLocalPath
+import com.microsoft.tfs.model.host.TfsModel
+import org.apache.log4j.Level
+import java.nio.file.Paths
+import kotlin.system.exitProcess
+
+fun main(args: Array<String>) {
+    println("Reactive TFS Client version 1.0")
+
+    val port = args.getOrNull(0)?.toIntOrNull()
+    if (port == null) {
+        printUsage()
+        exitProcess(1)
+    }
+
+    initializeApp(args)
+    runRdClient(port)
+}
+
+private fun printUsage() {
+    println("Usage:")
+    println("  <binary> <portNumber> [logDirectory]")
+    println()
+    println("- <portNumber>: port to connect to")
+    println("- [logDirectory]: path to the log directory (log file will be created automatically)")
+    println("- [logLevel]: log level (either ALL, DEBUG, INFO, WARN, ERROR, FATAL, OFF, or TRACE), INFO by default")
+}
+
+private fun initializeApp(args: Array<String>) {
+    val logDirectory = if (args.size > 1) Paths.get(args[1]) else null
+    val logLevel = if (args.size > 2) Level.toLevel(args[2]) else Level.INFO
+    Logging.initialize(logDirectory, logLevel)
+}
+
+private fun runRdClient(portNumber: Int) {
+    val logger = Logging.getLogger("Main")
+    logger.info { "Application initializing" }
+
+    val appLifetime = LifetimeDefinition()
+    val scheduler = SingleThreadScheduler(appLifetime, "com.microsoft.tfs.MainScheduler")
+    val socket = SocketWire.Client(appLifetime, scheduler, portNumber, "com.microsoft.tfs.MainSocketClient")
+    startSocketWatchdog(appLifetime, socket, logger)
```

Comment:
I wanted to add it, but it is non-trivial to do in Java 8, because it even has no good API to watch over other processes. Unfortunately, all the good stuff was added to Java 9 only.

What's implemented here is the best method of watching for parent process I could come up. I have tested it with various parent process termination cases, and it _seems_ to work well.

We'll improve it by adding heartbeat into the Rd protocol later.